drivers/clock_control/clock_control_nrf2_hsfll.c

```diff
@@ -0,0 +1,283 @@
+}
+
+static void hsfll_work_handler(struct k_work *work)
+{
+	struct hsfll_dev_data *dev_data =
+		CONTAINER_OF(work, struct hsfll_dev_data, clk_cfg.work);
+	uint32_t active_options;
+	uint8_t option_idx = ARRAY_SIZE(clock_options) - 1;
+	int rc;
+
+	active_options = clock_config_update_begin(work);
+	while (option_idx) {
+		if (active_options & BIT(option_idx)) {
+			break;
+		}
+		--option_idx;
+	}
+
+	rc = dvfs_service_handler_change_freq_setting(
+		clock_options[option_idx].setting);
+	if (rc < 0) {
+		clock_config_update_end(&dev_data->clk_cfg, rc);
+	} else {
+		k_timer_start(&dev_data->timer, UPDATE_TIMEOUT, K_NO_WAIT);
+	}
+}
+
+static struct onoff_manager *hsfll_find_mgr(const struct device *dev,
+					    const struct nrf_clock_spec *spec)
+{
+	struct hsfll_dev_data *dev_data = dev->data;
+	uint32_t frequency;
+
+	if (!spec) {
+		return &dev_data->clk_cfg.onoff[0].mgr;
+	}
+
+	if (spec->accuracy || spec->precision) {
+		LOG_ERR("invalid specification of accuracy or precision");
+		return NULL;
+	}
+
+	frequency = spec->frequency == NRF_CLOCK_CONTROL_FREQUENCY_MAX
+		  ? max_frequency
+		  : spec->frequency;
+
+	for (int i = 0; i < ARRAY_SIZE(clock_options); ++i) {
+		if (frequency > clock_options[i].frequency) {
+			continue;
+		}
+
+		return &dev_data->clk_cfg.onoff[i].mgr;
+	}
+
+	LOG_ERR("invalid frequency");
+	return NULL;
+}
+#endif /* CONFIG_NRFS_HAS_DVFS_SERVICE */
+
+static int api_request_hsfll(const struct device *dev,
+			     const struct nrf_clock_spec *spec,
+			     struct onoff_client *cli)
+{
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+	struct onoff_manager *mgr = hsfll_find_mgr(dev, spec);
+
+	if (mgr) {
+		return onoff_request(mgr, cli);
+	}
+
+	return -EINVAL;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+static int api_release_hsfll(const struct device *dev,
+			     const struct nrf_clock_spec *spec)
+{
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+	struct onoff_manager *mgr = hsfll_find_mgr(dev, spec);
+
+	if (mgr) {
+		return onoff_release(mgr);
+	}
+
+	return -EINVAL;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+static int api_cancel_or_release_hsfll(const struct device *dev,
+				       const struct nrf_clock_spec *spec,
+				       struct onoff_client *cli)
+{
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+	struct onoff_manager *mgr = hsfll_find_mgr(dev, spec);
+
+	if (mgr) {
+		return onoff_cancel_or_release(mgr, cli);
+	}
+
+	return -EINVAL;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+static int api_get_rate_hsfll(const struct device *dev,
+			      clock_control_subsys_t sys,
+			      uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sys);
+
+	*rate = nrf_hsfll_clkctrl_mult_get(NRF_HSFLL) * MHZ(16);
+
+	return 0;
+}
+
+static int hsfll_init(const struct device *dev)
+{
+	const struct hsfll_dev_config *dev_config = dev->config;
+	nrf_hsfll_trim_t trim = {
+		.vsup   = sys_read32(dev_config->trim_vsup),
+		.coarse = sys_read32(dev_config->trim_coarse),
+		.fine   = sys_read32(dev_config->trim_fine),
+	};
```

Comment:
Probably TCOEF is missing here. It was added in Lilium FP1. Potentially is not supported by HAL yet.